### ISSUE TITLE
Document search is not working anymore

### ISSUE THREAD

**Issue by aboydnw** [NASA-IMPACT/nasa-apt#500]
Here is the error in the console log when trying to conduct a search: 
![image](https://user-images.githubusercontent.com/92742765/167438546-e7c603c2-1700-4b2a-b398-f3e8f3790b40.png)

**Comment by leothomas** [comment]
Do you know if this occurring both staging and production? If not, which environment is it occurring in?

**Comment by aboydnw** [comment]
I pulled this screenshot from MCP. I'm not sure if it's happening in staging as well, since the entire documents page isn't accessible there.

**Comment by leothomas** [comment]
Update: the 502 error isn't being generated by the Lambda function or the API Gateway instance, it's generated by the Elasticsearch node itself. 

A few strange things to note about the Elasticsearch instance: 
### Last night: 
- Last night I couldn't load the instance info in the AWS console
- The Elasticsearch instance had no status/allocated memory in the Opensearch Domain overview tab: 
![image](https://user-images.githubusercontent.com/11858457/171427329-36b22a8c-327c-4105-b4af-7658eb43bfbe.png)
- Querying the Elasticsearch instance through the command line showed that the instance was "up and running as normal"
- I added logging to the Elasticsearch instance through the command line

### This morning: 
- The Elasticsearch instance now loads in the AWS console
- The Elasticsearch instance seems to be stuck in some sort of update processing state: 
<img width="821" alt="image" src="https://user-images.githubusercontent.com/11858457/171427810-aa7dd156-2cb9-4a63-988c-829dee952fb1.png">
(I'm not sure how this was triggered, perhaps by adding the logs last night?)

**Comment by leothomas** [comment]
Quick update on the search functionality in the **staging** instance: The AWS console still shows the Elasticsearch instance in an "in-progress" state, however the instance is now responsive. All of the documents that were indexed before the crash are lost, but I was able to index a new one and search it:
eg: querying keyword "ocean"
```bash
curl -X 'POST' \
  'https://0mrzuyq2e3.execute-api.us-east-1.amazonaws.com/v2/search' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer $AUTH_TOKEN' \
  -H 'Content-Type: application/json' \
  -d '{"query":{"bool":{"must":[{"multi_match":{"query":"oceans"}}],"filter":[]}},"highlight":{"fields":{"*":{}}}}'
``` 

result: 
```bash
{
  "took": 598,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 0.54655457,
    "hits": [
      {
        "_index": "atbd",
        "_type": "atbd",
        "_id": "19_v1",
        "_score": 0.54655457,
        "_source": {
           ...
        }
      ....
      ]
    }
```

**Comment by aboydnw** [comment]
Update: we would like to use option 2.1 outlined in https://github.com/NASA-IMPACT/nasa-apt/issues/509

**Comment by leothomas** [comment]
Update: after deploying an update to MCP which use the region from the [lambda's execution context](https://github.com/NASA-IMPACT/nasa-apt/pull/520/files#diff-05e026f8354f9aaf491d73b467109fe869541c8dbd3a8709f0e0f1eae7126fb2R20) to sign the Elasticsearch request, the document search still appears to be broken with an error message indicating that the signing credentials are scoped to `us-east-1` (when the ES instance is deployed to `us-east-1`).

**Comment by naomatheus** [comment]
[Proposed Solution 1](https://developmentseed.slack.com/files/T025QMQE3/F03MFD2DZML)

**Comment by naomatheus** [comment]
I've followed steps to [create a manual snapshot ](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/managedomains-snapshots.html) of the current elastic search index. However, the instance appears locked in "processing," and the Opensearch Dashboard is then inaccessible. I've updated the security configurations on the opensearch domain in DevSeed AWS, but the changes are not applying since the instance appears locked.

I was able to get through IAM roles and permissions required to create a snapshot of indices, but was not able to create the snapshot because the instance is locked.

There is an option to use the AWS dashboard to force through an update of this domain. And I believe in current indices may be lost during that process.

I am recommending both of the following: 
1. Processing the update from ElasticSearch to OpenSearch through the AWS console (risk losing indexes)
2. Deploy a new opensearch domain.
3. Configure the new opensearch domain to automate index snapshots and back them up to S3. 

I'll come back to this ticket after some discussion, and I've prepared the infrastructure code to deploy a new opensearch stack.

Here's a TLDR of what remains the same when migrating from ElasticSearch to OpenSearch.

## Upgrade to OpenSearch


The upgrade is automated and can be done automatically through the AWS management console. However, any changes made through the console would not be automatically reflected in CDK infrastructure code. So in this branch the infrastructure code has been updated. Notes are added on the following affected areas. Here are summaries of what will need to be changed and what can be left alone.

New API version
- OpenSearch continues to support both API versiions from ElasticSearch and OpenSearch.
  - Compatibility can  be referenced [here](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/configuration-api.html) 
Renamed instance types
- No action needed
Access policy changes
- IAM Policies
  - No actions needed. No IAM policies applied with CDK to ElasticSearch ->> OpenSearch resources.
- SCP Policies
  - The SCP policy applied to OpenSearch and many other services in `nasa-impact` allows all actions. No changes are required here. 
New resource types
- CDK Modifications
  - https://docs.aws.amazon.com/cdk/api/v1/docs/aws-elasticsearch-readme.html#migrating-to-opensearch
  - CDK infrastructure code in `stack.app` has been updated in this branch
Kibana renamed to OpenSearch Dashboards
Renamed CloudWatch metrics
- No CW metrics created in CDK. 
Billing and Cost Management console changes
  - https://docs.aws.amazon.com/opensearch-service/latest/developerguide/rename.html#rename-billing

New event format
- No existing EventBridge events that need [reformatting](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/rename.html#rename-billing).

What's staying the same?
- The following features and functionality, among others not listed, will remain the same:
  - Service principal (es.amazonaws.com)
  - Vendor code
  - Domain ARNs
  - Domain endpoints

**Comment by naomatheus** [comment]
**Consideration**
We don't have a way to handle permissions in opensearch, elasticsearch.
in APT, we do not have a way to limit the index search capability in line with user document access permissions. 
Adding permissions to the opensearch/elasticsearch layer.

Current solution:
Documents only get indexed when documents are published i.e. they're available to everyone.

**Comment by naomatheus** [comment]
Unblocked: There are no critical ATBDs/documents in APT in Prod accounts.
So there will be no need for migration steps to include creating a snapshot of current indices.
Those in development will also be lost when upgrading from ElasticSearch to OpenSearch.

**Comment by naomatheus** [comment]
Opensearch upgrade:
https://github.com/NASA-IMPACT/nasa-apt/compare/develop...open-search-upgrade-1

Ready to deploy pending review.

**Comment by naomatheus** [comment]
Remaining steps:

1. Change deployed open search endpoint from `https://0mrzuyq2e3.execute-api.us-east-1.amazonaws.com/v2/search` to `https://2t2dh1w620.execute-api.us-east-1.amazonaws.com/search`. 
2. ~~The new OpenSearch indices will be blank, and test ATBDs should be created in the stage environment.~~
3. ~~Search for contents of the new, indexed ATBDs~~

**Comment by naomatheus** [comment]
See this branch and PR thread for changes made.
Error state: ATBD documents displayed no documents

Current state:
[ATBD documents now displays documents](https://nasa-apt.surge.sh/documents)

Next steps:
Change deployed open search endpoint from https://0mrzuyq2e3.execute-api.us-east-1.amazonaws.com/v2/search to https://2t2dh1w620.execute-api.us-east-1.amazonaws.com/search

**Comment by naomatheus** [comment]
Upcoming changes to Backend API for Opensearch compatibility
- Replace requests_aws4auth with opensearch-py/opensearchpy
   - requests_aws4auth does not currently support AWS Opensearch service and is maintained outside of the Opensearch organization
   - opensearch-py is supported and maintained by Opensearch.org, is opensource licensed, and is compatible with Opensearch v1.0.1, our current version of Opensearch
   - Affected modules are `elasticsearch.py` `app/api/v2/elasticsearch` at  and `elasticsearch.py` at `app/search/elasticsearch.py`

PR forthcoming, in the meantime see `app/api/v2/elasticsearch.aws_auth`. AWS auth function returns an object used to authorize post requests to Elasticsearch domain. In the update, the client must be used directly with queries. (reason: [opensearch-py is low level client for now](https://pypi.org/project/opensearch-py/#getting-started))

- Setup.py is deprecated (?)
  - This should be replaced with a shell script that uses pip for package management
  -  `setup.py install is deprecated`

**Comment by bwbaker1** [comment]
@naomatheus @wrynearson @aboydnw @deborahUAH After testing open search on staging, it does work, but there are still a few issues. 

(1) A search term must be entered.  We would like for a user to be able to search for any ATBDs published in a specific year. For example, leave the search term blank and just search for all ATBDs published in 2022. 

![apt_1](https://user-images.githubusercontent.com/56319064/197024870-47b4bd3b-6231-4e89-bb13-633868f9ac73.jpeg)

(2) When a search term is included, the year must still be "All." The image below shows no document matching "contributor" when searching using the year as 2022.  This particular ATBD was published in 2022. 

![apt_2](https://user-images.githubusercontent.com/56319064/197025656-fcf6b08e-8dfd-437d-9fde-b9cdca57bf93.jpeg)

Below is the search results with the search term "contributor" and the year is set to "all." 

![apt_3](https://user-images.githubusercontent.com/56319064/197025926-cad09194-9935-4c77-964b-297956948b39.jpeg)

**Comment by wrynearson** [comment]
Thanks @bwbaker1. I think we want to prioritize having a *functional* production environment quickly. To do so, would you and @deborahUAH agree to push our *current* implementation of this ticket from staging to production?

**Comment by bwbaker1** [comment]
@wrynearson My opinion is that it is functional enough to go ahead and push to PROD.  Then we can get the DEMO ATBD public since that is a big deal.  The other parts of this ticket can be fixed afterward.  But this is @deborahUAH decision.

**Comment by wrynearson** [comment]
OK, thanks for your quick reply. If @deborahUAH approves this approach, please let us know here, then @naomatheus could push to production when he's ready.

**Comment by bwbaker1** [comment]
@wrynearson @naomatheus I discussed this with Deborah and she gave the go ahead to push this to production.

**Comment by naomatheus** [comment]
@bwbaker1  If you have a chance could you go through the process of creating a new document, approving, and publishing the new document in MCP (https://www.earthdata.nasa.gov)? The document should have at least some minimal text.

**Comment by deborahUAH** [comment]
@naomatheus  do you mean in production (not staging)??  If I do this (I just created a new fake ATBD) and I approve and publish it in MCP then it will be visible to all users, and we will not be able to delete it.  So.... not sure I want to do that!

**Comment by wrynearson** [comment]
@deborahUAH , @naomatheus was suggesting this to address #591.

Instead, a demo ATBD can be created in production. This demo ATBD will then be "indexed" once it is published. We don't need a "test" ATBD.

@naomatheus, please confirm that @deborahUAH and @bwbaker1 can create a demo ATBD before you've fixed #545 .

**Comment by naomatheus** [comment]
Yes, @deborahUAH @bwbaker1  you can create a demo ATBD, publish it, and then the document will be indexed and searchable once it is a published. This can occur before #545  is resolved.

 I've gone through the publication process with an ATBD to test the functionality, it is named after a movie title (burn after reading) and has very little content. I was not aware that an ATBD could not be deleted by the owner/curator roles once published.

If that "test atbd" is going to cause an issue I can find out how to remove it.

**Comment by bwbaker1** [comment]
@naomatheus Maybe I am wrong in my thinking, but I think we need to be able to download the DEMO ATBD before it gets published [#545](https://github.com/NASA-IMPACT/nasa-apt/issues/545#issue-1326347943).  We need to ensure everything looks correct since it will be public and we can't delete it once it is public (so it needs to look good). @deborahUAH Correct me if I am wrong here.

**Comment by wrynearson** [comment]
Sorry for this confusion. I was able to delete the "test" ATBD with my account will@developmentseed.org. This seems to be a bug from https://github.com/NASA-IMPACT/nasa-apt/issues/548#issuecomment-120859528 that will be fixed in the next PI. @bwbaker1 please confirm that you can no longer see this test ATBD.

**Comment by deborahUAH** [comment]
Brad is correct.  We need to download the real DEMO ATBD that @bwbaker1 has created in prod to see that it looks ok before we publish it in prod (because at that point it will be visible for good).  Publishing in MCP Prod is not a testing space.  This is why we needed a staging space on MCP and I thought one was created months ago?? (there was a ticket for it somewhere - please add # if you find it).   

The  order in MCP Prod must be 1) download the DEMO ATBD PDF and check that it looks good (enough), _then_ publish it, _then_ check the ability to see it when searching.   
@wrynearson  @naomatheus

**Comment by bwbaker1** [comment]
@wrynearson @deborahUAH I just checked and I no longer see the test ATBD on PROD. So it seems you were successful in deleting that document.

**Comment by deborahUAH** [comment]
see my clarification on deletions on ticket #548

**Comment by naomatheus** [comment]
Hey @wrynearson . Are we able to move this issue to done for now? 
I'd suggest splitting off another issue if there's more to do here for this specific issue.

**Comment by wrynearson** [comment]
Hey @naomatheus , sorry I just saw this.

@deborahUAH is the owner of this repo, so I'm not able to close tickets. She treats `done` as it's deploy to prod and tested. Once their team writes a demo ATBD, downloads it while it's in draft, then publishes it and makes sure it's indexed, then they will close it.

**Comment by naomatheus** [comment]
Gotcha. Noted @wrynearson

**Comment by deborahUAH** [comment]
awaiting more than 1 document so this can be tested on Production.

**Comment by wrynearson** [comment]
@deborahUAH @bwbaker1 now that we have Document PDFs working in production, this ticket is unblocked. I've marked it for your review.

**Comment by bwbaker1** [comment]
Note: we can't review until we have published documents.

**Comment by bwbaker1** [comment]
@wrynearson Search is no longer working on staging. It does not provide error or anything. 


https://user-images.githubusercontent.com/56319064/230947193-c9050799-0645-47be-8286-478abe811c4b.mov

**Comment by wrynearson** [comment]
Thanks for flagging this @bwbaker1 

@naomatheus and maybe @thenav56 - could you look into this? I'm getting `500` errors in the console, so this might be an AWS issue.

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/25919630/231152598-f34194f5-0fb5-423a-8916-e0b16fc2ca05.png">

**Comment by wrynearson** [comment]
@sunu @thenav56, do we need to make any further updates on production to make sure that this ticket can be close (e.g. https://github.com/NASA-IMPACT/nasa-apt/pull/706)?

**Comment by wrynearson** [comment]
@thenav56 @sunu @batpad currently, no results are appearing on staging for document search. There is no error, but results that we think should appear aren't.

cc @bwbaker1

**Comment by bwbaker1** [comment]
⬆️  Here is a quick screen recording of searching. 

https://user-images.githubusercontent.com/56319064/235491001-3651154a-ca58-4af2-a030-b4c9ec7fcdad.mov

**Comment by wrynearson** [comment]
@bwbaker1 the issue is that public documents that were made public _before_ we implemented https://github.com/NASA-IMPACT/nasa-apt/pull/699 are not "indexed" (i.e., not searchable).

We made two documents on staging that are searchable now:

https://user-images.githubusercontent.com/25919630/235601485-542cb4a6-a3f1-4147-9ea9-2f3f0dc3b100.mov

This won't be an issue on production because we don't have any published documents there yet, so once the first document is published, it will be indexed. However, we will go back and re-index the existing published documents on staging to not cause any further confusion.

cc @thenav56 @sunu

**Comment by bwbaker1** [comment]
@wrynearson This makes sense. I remember this happening awhile back when search was updated/fixed.
 @thenav56 @sunu @batpad Thanks for getting this solved so quickly!

**Comment by wrynearson** [comment]
@thenav56 should we wait to push this to production until https://github.com/NASA-IMPACT/nasa-apt/issues/718 is ready? cc @sunu @batpad

**Comment by sunu** [comment]
@wrynearson IMO there is no need to wait since the production instance doesn't have any published ATBDs anyway

**Comment by thenav56** [comment]
@wrynearson Yes, as @sunu said we are good to push this to production.
https://github.com/NASA-IMPACT/nasa-apt/issues/718 can wait as it will add a feature to fix ES issues if there are any in the future.
